### PR TITLE
Drop patch for profile_io_data.cc and use chromium_src instead

### PR DESCRIPTION
**Pull request by mariospr**
The idea of this patch was simply to make sure that the brave://
scheme gets considered by ProfileIOData::IsHandledProtocol() along
with other schemes, but we can achieve the same goal just using
chromium_src overrides, so let's do that instead.

Fixes https://github.com/brave/brave-browser/issues/14238

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)
- [ ] Requested a security/privacy review as needed

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

See Test plan from https://github.com/brave/brave-core/pull/1196

### CODE REVIEW

File: chromium_src/chrome/browser/profiles/profile_io_data.cc
Comment (iefremov) [inline chromium_src/chrome/browser/profiles/profile_io_data.cc:21]: don't quite follow what this is for

Comment (mariospr) [reply chromium_src/chrome/browser/profiles/profile_io_data.cc:21]: `IsHandledProtocol()` is referenced from the body of `IsHandledURL()` in this `profile_io_data.cc` file, and thus the renaming via `#define IsHandledProtocol IsHandledProtocol_ChromiumImpl` done above this lines also changes that line inside `IsHandledURL()`'s body.

Therefore, I understand that we need to provide our own version of `IsHandledURL()` as well to make sure that we get our own version of `IsHandledProtocol()` called when invoked via `IsHandledURL()`, no?

Comment (iefremov) [reply chromium_src/chrome/browser/profiles/profile_io_data.cc:21]: gotcha, I just didn't look into what's happening in `IsHandledURL()`. Probably worth adding a comment

**Review by iefremov — APPROVED** [review]

File: chromium_src/chrome/browser/profiles/profile_io_data.cc
Comment (mariospr) [reply chromium_src/chrome/browser/profiles/profile_io_data.cc:21]: No problem, comment added. Will merge once CI finishes running again. Thanks!

**Comment by mariospr** [comment]
@brave/patch-reviewers Looks like I still need a review from you, PTAL when you have a moment, thanks!

File: chromium_src/chrome/browser/profiles/profile_io_data.cc
Comment (mkarolin) [reply chromium_src/chrome/browser/profiles/profile_io_data.cc:25]: This method seems to be missing 
<pre>
  if (!url.is_valid()) {
    // We handle error cases.
    return true;
  }
</pre>

I am a bit hesitant about replacing methods in this way: if the original method is further modified we'd never know.

Comment (mariospr) [reply chromium_src/chrome/browser/profiles/profile_io_data.cc:25]: > This method seems to be missing
>
>   if (!url.is_valid()) {
>     // We handle error cases.
>     return true;
>   }

Oops! True, added now.

> I am a bit hesitant about replacing methods in this way: if the original method is further modified we'd never know.

I share the same concern to be honest. In the other hand, I think this risk already exists in other overrides since renaming a symbol via define will change not just the name of the method in its definition, but also any callpoint inside the same .cc file. That is, imagine another that we already have where we rename an internal method that was originally not called from any other point in that .cc file, and then upstream decides to reference it at some point from another method in that file: our override might not be correct now since that newly added callpoint would now be invoking the unmodified logic, not ours... and we don't have an easy way to detect that, I think?

Now back to this particular case, I did check what `git blame` had to say about `IsHandledURL()` and found this:
```
672c8c1e10e9b (dhollowa@chromium.org      2013-03-07 12:30:06 +0000 279) // static
a8c1e745a571a (willchan@chromium.org      2011-05-14 06:17:07 +0000 280) bool ProfileIOData::IsHandledURL(const GURL& url) {
a8c1e745a571a (willchan@chromium.org      2011-05-14 06:17:07 +0000 281)   if (!url.is_valid()) {
a8c1e745a571a (willchan@chromium.org      2011-05-14 06:17:07 +0000 282)     // We handle error cases.
a8c1e745a571a (willchan@chromium.org      2011-05-14 06:17:07 +0000 283)     return true;
a8c1e745a571a (willchan@chromium.org      2011-05-14 06:17:07 +0000 284)   }
```
That is, it seems that the **last time this `IsHandledURL()` method was modified was 10 years ago, so perhaps the risk in this particular case is low.

@mkarolin WDYT?

PS: FWIW, I'm not pushing for this to be done, I'm happy to go back to having a .patch and re-think how to get rid about it (or keep it), just mentioning this because it was the reasoning I followed

Comment (mkarolin) [reply chromium_src/chrome/browser/profiles/profile_io_data.cc:25]: Agree, not a strong concern.

**Review by mkarolin — APPROVED** [review]
++

**Comment by mariospr** [comment]
The failure in the post-init CI bot is unrelated to this PR and this can be safely merged now, so doing it now. Thanks!

**Pull request merged**
